bench/workgen/workgen.cxx

```diff
@@ -2381,6 +2410,33 @@ Workload::get_tables()
     return uris;
 }
 
+const std::vector<std::string>
+Workload::garbage_collection()
+{
+    ContextInternal *icontext = _context->_internal;
+    const std::lock_guard<std::mutex> lock(*icontext->_dyn_mutex);
+    std::vector<std::string> uris;
+
+    for (size_t i = 0; i != icontext->_dyn_tables_delete.size();) {
+        // The table might still be in use.
+        const std::string uri(icontext->_dyn_tables_delete.at(i));
+        if (icontext->_dyn_table_in_use[uri] != 0) {
+            ++i;
+            continue;
+        }
+
+        // Delete all local data related to the table.
+        icontext->_dyn_table_in_use.erase(uri);
+        icontext->_dyn_tables_delete.erase(icontext->_dyn_tables_delete.begin() + i);
+        tint_t tint = icontext->_dyn_tint.at(uri);
+        icontext->_dyn_tint.erase(uri);
+        icontext->_dyn_table_names.erase(tint);
+        // TODO - _dyn_table_runtime
```

Comment:
This one is tricky. It is used this way:
```
recno = workgen_atomic_add64(&_icontext->_dyn_table_runtime.at(tint)._max_recno, 1);
```
If I delete the entry, the vector will be reduced in size and this will shift everything and the statement just above might become wrong.
Seems like a `std::map` would do a better job, thoughts?

Reply:
It looks to me like the use you cite needs to also hold the lock.  I don't think std::map is better.  Like std::vector, it is not built for concurrent access (where at least one access is a writer).

I think the best approach is to hold a lock for really every access - read or write.  Don't try to cut corners for performance.  Use WT-10309 to look at performance, I think the shared mutex will help.

Reply:
@ddanderson:
> It looks to me like the use you cite needs to also hold the lock.

This was in the end added in the other PR, [here](https://github.com/wiredtiger/wiredtiger/pull/8573/files#diff-fded55a163215494e7abd497d1b8ee130da02cc03c7f8da23bf54dce10440f31R1019).

> I don't think std::map is better. Like std::vector, it is not built for concurrent access (where at least one access is a writer).

I meant in terms of not wasting memory. Right now I cannot change the vector size as I am afraid to shift everything. With a map, I could delete the invalid key and free the memory that is associated with it.

Reply:
I have added a comment here 75d660b469ecc87000355ba609dcca029e66ede9. But I want to give it a shot with a map, I think it's better and less prone to errors when using the data structure.
Instead of having `std::vector<TableRuntime> _dyn_table_runtime`, we would have `std::map<std::string, TableRuntime> _dyn_table_runtime;`. This way we can simply use the table uri to access the `TableRuntime` object` and delete the entry when no longer needed.

Reply:
I have created a map in this commit 20a506a02aecad1070a0030514d1438d40dd4eda.
I think it is better as I can free the no longer used key [here](https://github.com/wiredtiger/wiredtiger/pull/8579/commits/20a506a02aecad1070a0030514d1438d40dd4eda#diff-fded55a163215494e7abd497d1b8ee130da02cc03c7f8da23bf54dce10440f31R2433).